src/v/kafka/server/request_context.h

```diff
@@ -272,6 +237,69 @@ class request_context {
         usage_mgr().add_bytes_sent(response_size - internal_bytes_sent);
     }
 
+    template<typename ResponseType>
+    ss::future<ResponseType> log_response(ResponseType r) {
+        auto msg = fmt::format(
+          "[{}:{}] sending {}:{} for {}, response {}",
+          _conn->client_host(),
+          _conn->client_port(),
+          ResponseType::api_type::key,
+          ResponseType::api_type::name,
+          _header.client_id,
+          r);
+
+        return r.data.errored().then(
+          [msg, r = std::move(r)](auto errored) mutable {
+              if (errored) {
+                  thread_local static ss::logger::rate_limit rate(
+                    std::chrono::seconds(10));
+                  klog.log(ss::log_level::debug, rate, "{}", msg);
+              } else {
+                  vlog(klog.trace, "{}", msg);
+              }
+              return ss::make_ready_future<ResponseType>(std::move(r));
+          });
+    }
+
+    template<typename ResponseType>
+    requires requires(
+      ResponseType r, protocol::encoder& writer, api_version version) {
+        { r.encode(writer, version) } -> std::same_as<void>;
+    }
+    ss::future<response_ptr> process_response(ResponseType r) {
+        /// Many responses contain a throttle_time_ms field, to prevent each
+        /// handler from manually having to set this value, it can be done in
+        /// one place here, with this concept check
+        if constexpr (has_throttle_time_ms<ResponseType>) {
+            /// Allow request handlers to override the throttle response, if
+            /// multiple throttles detected, choose larger of the two
+            r.data.throttle_time_ms = std::max(
+              r.data.throttle_time_ms, throttle_delay_ms());
+        }
+
+        /// KIP-511 bumps api_versions_request/response to 3, past the first
+        /// supported flex version for this API, and makes an exception
+        /// that there will be no tags in the response header.
+        auto is_flexible = flex_enabled(header().is_flexible());
```

Comment:
Converted the PR to draft because manual tests reveal a `stack-use-after-return` runtime error here. Need to find a way around this.

Reply:
`stack-use-after-return` because when we do
```cpp
     co_return ctx.respond(std::move(resp));
```

the instance of `request_context` is destroyed.

Reply:
FYI: I am simply taking notes here. I am not looking for feedback.